### PR TITLE
Check spots available

### PR DESCRIPTION
**Pull request by Amotul-raheem**
this endpoint checks the data base to see pending bookings. it compares the check in and check out dates to those in the database to see if doesn't fall in the range of any pending bookings. And it returns list of unavailable parking spots.

### CODE REVIEW

File: routes/SpotsAvailable.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    for (let i = 0; i < pendingBookings.length; i++) {
+        let x = (pendingBookings[i].space_id);
+        let y = (pendingBookings[i].check_in);
+        let z = (pendingBookings[i].check_out);
+
+        const dates = getDates(new Date(y), new Date(z))
+        dates.forEach(function (date) {
+            // console.log(date)
+            if (date === checkOut || date === checkIn) {
+                unavailableSpots.push(x)
```
Comment (Amotul-raheem) [inline routes/SpotsAvailable.js]: I'm having issue with this if statement. Everything under it is nit implementing and i think has somethin to do with the date format. Still trying to resolve it.

File: routes/SpotsAvailable.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        })
+    }
+    console.log(unavailableSpots)
+    res.send("working")
+
+
+})
+
+
+// Usage
```
Comment (Amotul-raheem) [inline routes/SpotsAvailable.js]: i haven't cleaned up the code yet. Will get rid of these soon.

File: routes/BookingRouter.js [hunk 3]
```diff
@@ -41,22 +41,7 @@ bookingRouter.post("/booking", authVerify, async (req, res) => {
         res.status(500).send(error);
     }
 })
-bookingRouter.post("/isSpotAvailable", async (req, res) => {
```
Comment (Amotul-raheem) [inline routes/BookingRouter.js]: This was my previous approach. couldn't get it to work

File: app.js [hunk 4]
```diff
@@ -18,7 +19,7 @@ app.use(bodyParser.urlencoded({limit: "30mb", extended: true}));
 const PORT = process.env.PORT
 
 app.use("/api", bookingRouter);
-// app.use("/api", passwordRouter)
+app.use("/api", spotsAvailableRouter);
```
Comment (folaraz) [inline app.js]: change name to ParkSpotRetreiver

File: routes/SpotsAvailable.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    let unavailableSpots = []
+    const pendingBookings = await Booking.find({booking_status: BOOKING_STATUS.PENDING})
+    var dt = new Date("30 July 2010 15:05 UTC");
+    document.write(dt.toISOString());
+
+    const checkIn = req.body.check_in
+    const checkOut = req.body.check_out
+
+    for (let i = 0; i < pendingBookings.length; i++) {
+        let x = (pendingBookings[i].space_id);
```
Comment (folaraz) [inline routes/SpotsAvailable.js]: 🙄 whats x?

File: app.js [hunk 6]
```diff
@@ -18,6 +19,7 @@ app.use(bodyParser.urlencoded({limit: "30mb", extended: true}));
 const PORT = process.env.PORT
 
 app.use("/api", bookingRouter);
+app.use("/api", totalSpotsRouter);
```
Comment (folaraz) [inline app.js]: change it to ParkSpotsRouter, that name doesn't make sense

File: routes/SpotsRouter.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
+import express from "express";
+import moment from "moment";
+import {ParkSpace} from "../models/ParkSpace.js";
+import {BOOKING_STATUS} from "../utils/BookingStatus.js";
+import Booking from "../models/Booking.js";
+
+const totalSpotsRouter = express.Router();
```
Comment (folaraz) [inline routes/SpotsRouter.js]: Change to ParkSpotRouter, don't forget

File: routes/SpotsRouter.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+totalSpotsRouter.post("/isSpotAvailable", async (req, res) => {
+    try {
+        const reqCheckIn = moment(req.body.check_in).format()
+        const reqCheckOut = moment(req.body.check_out).format()
+
+        let totalSpots = {}
```
Comment (folaraz) [inline routes/SpotsRouter.js]: call it allParkSpots instead

File: routes/SpotsRouter.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+        const pendingAndInuseBookings = await Booking.find({$or: [{booking_status: BOOKING_STATUS.PENDING}, {booking_status: BOOKING_STATUS.INUSE}]})
+        const fulfilledAndCancelledBookings = await Booking.find({$or: [{booking_status: BOOKING_STATUS.FULFILLED}, {booking_status: BOOKING_STATUS.CANCELLED}]})
+
+        let availableSpots = fulfilledAndCancelledBookings.map(getAvailableSpaceId)
```
Comment (folaraz) [inline routes/SpotsRouter.js]: it's returning spotIds not spots. So say availabeSpotIds.


you don't need a new getAvailableSpaceIds function. You can easily, do `let availableSpotIds = fulfilledAndCancelledBookings.map(booking => booking.space_id);`

Therefore, you don't  need the function you wrote below. Also, the function doesn't look right to me.

File: routes/SpotsRouter.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        let unavailableSpots = pendingAndInuseBookings.map(getUnavailableSpaceId)
```
Comment (folaraz) [inline routes/SpotsRouter.js]: same here, unavailabeSpotIds

File: routes/SpotsRouter.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        let unavailableParkSpaces = await ParkSpace.find({_id: {$in: unavailableSpots}}, {space_name: 1, _id: false});
+        let availableParkSpaces = await ParkSpace.find({_id: {$in: availableSpots}}, {space_name: 1, _id: false});
+
+        function getAvailableSpaceId(booking) {
+            let spaces = []
+            spaces.push(booking.space_id)
+            return spaces
+        }
+
+        function getUnavailableSpaceId(booking) {
```
Comment (folaraz) [inline routes/SpotsRouter.js]: This won't work. You pushing  to an array while looping through the pendingAndInUseBookings, doesn't make sense.


Use the example below to fix it and make a callback for line 20
`const euro = [29.76, 41.85, 46.5];

const above30 = euro.reduce((total, amount) => {
  if (amount > 30) {
    total.push(amount);
  }
  return total;
}, []);

above30 // [ 41.85, 46.5 ]`

File: routes/SpotsRouter.js [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+                spaces.push(booking.space_id)
+            }
+            return spaces
+        }
+
+        totalSpots.unavailabe = unavailableParkSpaces
+        totalSpots.available = availableParkSpaces
+
+        console.log(totalSpots)
+        res.status(200).send("Parkspaces updated successfully")
```
Comment (folaraz) [inline routes/SpotsRouter.js]: are we not meant to return all the spots?

File: routes/SpotsRouter.js [hunk 13]
```diff
@@ -0,0 +1,62 @@
+import express from "express";
+import moment from "moment";
+import {ParkSpace} from "../models/ParkSpace.js";
+import {BOOKING_STATUS} from "../utils/BookingStatus.js";
+import Booking from "../models/Booking.js";
+
+const totalSpotsRouter = express.Router();
+
+totalSpotsRouter.post("/isSpotAvailable", async (req, res) => {
```
Comment (folaraz) [inline routes/SpotsRouter.js]: this endpoint should not be called isSpotAvailable, instead getParkSpots

File: utils/BookingStatus.js [hunk 14]
```diff
@@ -1,5 +1,6 @@
 export const BOOKING_STATUS = {
     PENDING: "PENDING",
     FULFILLED: "FULFILLED",
-    CANCELLED: "CANCELLED"
+    CANCELLED: "CANCELLED",
+    INUSE: "IN-USE"
```
Comment (folaraz) [inline utils/BookingStatus.js]: change from IN-USE to IN_USE

**Review by folaraz — CHANGES_REQUESTED** [review]

**Review by folaraz — APPROVED** [review]

**Review by folaraz — APPROVED** [review]

**Review by folaraz — APPROVED** [review]

**Pull request merged**